### DISCUSSION TITLE
Edge classification/regression using Pytorch Geometric?

### DISCUSSION THREAD

**Discussion by antoni4040** [pyg-team/pytorch_geometric#3554] [Q&A] [answered]
Hello. I would like to do edge regression in Pytorch Geometric. I've only found information about it in DGL. ([link](https://docs.dgl.ai/en/0.6.x/guide/training-edge.html))
They basically suggest using a GNN to calculate a hidden embedding for each node and then take the dot product between nodes connected by edges. This is quite simple in DGL, but I need some help to do it in Pytorch Geometric, as I'm not very proficient (yet).
(The edge-wise dot product, not the GNNs obviously.)
I think edge classification/regression can solve quite a few problems, so any other information about it is welcome. Thanks! 😄

**Comment by rusty1s** [comment, marked as answer]
That is actually quite simple in PyG as well:
```python
x = ...  # Node feature matrix: [num_nodes, num_features]
edge_index = ... # Edge indices: [2, num_edges]

src, dst = edge_index
score = (x[src] * x[dst]).sum(dim=-1)

loss = F.cross_entropy(score, data.edge_label)  # Classification
loss = F.mse_loss(score, data.edge_value)  # Regression
```

**Reply by antoni4040** [reply]
Wow, that was much simpler than expected, thanks!

**Reply by michalisfrangos** [reply]
> That is actually quite simple in PyG as well:
> 
> ```python
> x = ...  # Node feature matrix: [num_nodes, num_features]
> edge_index = ... # Edge indices: [2, num_edges]
> 
> src, dst = edge_index
> score = (x[src] * x[dst]).sum(dim=-1)
> 
> loss = F.cross_entropy(score, data.edge_label)  # Classification
> loss = F.mse_loss(score, data.edge_value)  # Regression
> ```

Is it possible to open this discussion please ?

**Comment by michalisfrangos** [comment]
Hi rusty1s, 

Wouldn't this work only if there was a single edge from node to node? What happens if there were multiple edges between two nodes?

One approach I can think of is to  use edge embeddings. If we could extract the edge embeddings (for example in GATConv) then we could build a classification layer given the edge embedding.  Any ideas how to extract the edge embeddings from the GATConv layer?  

thanks

**Reply by rusty1s** [reply]
What do you mean with multiple edges? Typed edges? If you want to integrate edge embeddings, you can do that in a similar fashion as described [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/rgcn_link_pred.py#L47-L59).

**Reply by michalisfrangos** [reply]
Hi Matthias,

With multiple edges between two nodes, I do not mean different type edges. I mean same type edges with different message. For example person A (with node embedding h_A) sends to person B (with node embedding h_B) N messages represented as N edges with a different message (which can be represented with an embedding). 

Thanks ,
Michalis

**Reply by rusty1s** [reply]
How do you generate different messages for the same source and destination node features? Can't you apply the same procedure for your decoder as well?

**Reply by michalisfrangos** [reply]
Thanks Matthias for the response . 

(Perhaps I need to dive deeper and  implement from scratch , but at the moment I'm trying to learn how to use PyG flexibly ... still on a learning curve)

To clarify with an example: From airport A we can have multiple flights to the airport B in a day (multiple edges). The embedding of each edge can incorporate information of the plane , etc . 
Perhaps it's easier to refer to the book "graph representation learning , William L. Hamilton 2020 for the generalised message passing:

![image](https://user-images.githubusercontent.com/17205240/145722914-ebba52e2-f60a-4bed-b1a8-fd01c7596900.png)

For example at GATConv , we have access to the embeddings in equation 5.48, but not for the embedding in equation 5.46 (even though the edge embeddings are present, following the generalised message passing attached). If we had access to the embeddings of the edges in  5.46, we could be building a classifier for multiple   edges. 

Thanks 
M

**Reply by rusty1s** [reply]
The only way to currently extract the messages of a given GNN operator is to utilize our [hooks](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.message_passing.MessagePassing.register_propagate_forward_hook). Nonetheless, creating messages as in 5.46 is straightforward by following the procedure mentioned above, e.g.:
```python
src, dst = edge_index
message(x[src], x[dst], edge_attr)
```
Importantly, you only get different messages for the same node pair if the edge embedding differs. In your decoder, this may look as follows:
```python
x = ...  # Node feature matrix: [num_nodes, num_features]
edge_index = ... # Edge indices: [2, num_edges]

src, dst = edge_index
score = (x[src] * x[dst] * edge_attr).sum(dim=-1)

loss = F.cross_entropy(score, data.edge_label)  # Classification
loss = F.mse_loss(score, data.edge_value)  # Regression
```

**Reply by michalisfrangos** [reply]
```

Importantly, you only get different messages for the same node pair if the edge embedding differs.
```
Thanks Matthias, I think I get it now.. different edge_attr for a pair node will give a different score. I'll try this out and see how it goes. Thanks again for your responses !

**Reply by vikasbammidi** [reply]
Hi @michalisfrangos do you have any working complete code sample  on edge cassification,
